File: cpp_solutions/LinkedList/Merge_K_sorted_Lists.cpp
/*
* Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.
* 
**/
// Divide and Conquer, T = O(NlogN)
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
       if(lists.empty()) return NULL;

       int n = lists.size();
       while(n>1){
       	int k = (n+1)/2;
       	for(int i=0; i< n/2; i++){
       		lists[i] = mergeTwoLists(lists[i], lists[i+k]);
       	}
       	n= k;
       }
       return lists[0];
    }

     ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
     	ListNode* dummy = new ListNode(0);
     	ListNode* tmp = dummy;
     	while(l1 != NULL && l2 != NULL){
     		if(l1->val < l2->val){
     			tmp->next = l1;
     			l1 = l1->next;
     		}
     		else{
     			tmp->next = l2;
     			l2 = l2->next;
     		}
     		tmp = tmp->next;
     	}
     	if(l1 != NULL) tmp->next = l1;
     	else tmp->next = l2;
     	return dummy->next;
     }
};

File: python_solutions/LinkedList/Merge_K_Sorted_Lists.py
from heapq import heappush, heappop
class Solution:
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        if lists is None:
            return None
        
        pHead = ListNode(-1)
        pCur = pHead
        heap = []
        for ln in lists:
            if ln:
                self.pushHeap(heap, ln)
        
        while pCur:
            curNode = heappop(heap)[1]
            pCur.next = curNode
            pCur = pCur.next
            if pCur.next:
                self.pushHead(heap, pCur.next)
        return pHead.next
                
    
    
    def pushHeap(self, heap, node):
        heappush(heap, (node.val, node))


# divide and conquer
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        if not len(lists):
            return None
        return self.MergeHelper(lists, 0, len(lists)-1)
    
    def MergeHelper(self, lists, start, end):
        if start == end:
            return lists[start]
        mid = start + int((end - start)/2)
        left = self.MergeHelper(lists, start, mid)
        right = self.MergeHelper(lists, mid+1, end)
        return self.MergeTwoLists(left, right)
    
    def MergeTwoLists(self, l1, l2):
        if not l1:
            return l2
        if not l2:
            return l1
        
        pHead = ListNode(-1)
        pCur = pHead
        while l1 and l2:
            if l1.val <= l2.val:
                pCur.next = l1
                l1 = l1.next
            else:
                pCur.next = l2
                l2 = l2.next
            pCur = pCur.next
        if l1 is not None:
            pCur.next = l1
        else:
            pCur.next = l2
        
        return pHead.next

               